In [31]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import roc_auc_score, accuracy_score, confusion_matrix
%matplotlib inline
from matplotlib import pyplot as plt
import seaborn as sns

In [32]:
train_df = pd.read_csv('../../Desktop/titanic_train.csv')
test_df = pd.read_csv('../../Desktop/titanic_test.csv')

In [33]:
y = train_df['Survived']

In [34]:
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [35]:
train_df.describe(include='all')

PassengerId    Survived      Pclass  \
count    891.000000  891.000000  891.000000   
unique          NaN         NaN         NaN   
top             NaN         NaN         NaN   
freq            NaN         NaN         NaN   
mean     446.000000    0.383838    2.308642   
std      257.353842    0.486592    0.836071   
min        1.000000    0.000000    1.000000   
25%      223.500000    0.000000    2.000000   
50%      446.000000    0.000000    3.000000   
75%      668.500000    1.000000    3.000000   
max      891.000000    1.000000    3.000000   

                                      Name   Sex         Age       SibSp  \
count                                  891   891  714.000000  891.000000   
unique                                 891     2         NaN         NaN   
top     Pickard, Mr. Berk (Berk Trembisky)  male         NaN         NaN   
freq                                     1   577         NaN         NaN   
mean                                   NaN   NaN   29.699118    0.523008   
std                                    NaN   NaN   14.526497    1.102743   
min                                    NaN   NaN    0.420000    0.000000   
25%                                    NaN   NaN   20.125000    0.000000   
50%                                    NaN   NaN   28.000000    0.000000   
75%                                    NaN   NaN   38.000000    1.000000   
max                                    NaN   NaN   80.000000    8.000000   

             Parch Ticket        Fare    Cabin Embarked  
count   891.000000    891  891.000000      204      889  
unique         NaN    681         NaN      147        3  
top            NaN   1601         NaN  B96 B98        S  
freq           NaN      7         NaN        4      644  
mean      0.381594    NaN   32.204208      NaN      NaN  
std       0.806057    NaN   49.693429      NaN      NaN  
min       0.000000    NaN    0.000000      NaN      NaN  
25%       0.000000    NaN    7.910400      NaN      NaN  
50%       0.000000    NaN   14.454200      NaN      NaN  
75%       0.000000    NaN   31.000000      NaN      NaN  
max       6.000000    NaN  512.329200      NaN      NaN

In [36]:
train_df['Age'].fillna(train_df['Age'].median(), inplace=True)
test_df['Age'].fillna(train_df['Age'].median(), inplace=True)
train_df['Embarked'].fillna('S', inplace=True)
test_df['Fare'].fillna(train_df['Fare'].median(), inplace=True)

In [37]:
train_df = pd.concat([train_df, pd.get_dummies(train_df['Pclass'], 
                                               prefix="PClass"),
                      pd.get_dummies(train_df['Sex'], prefix="Sex"),
                      pd.get_dummies(train_df['SibSp'], prefix="SibSp"),
                      pd.get_dummies(train_df['Parch'], prefix="Parch"),
                     pd.get_dummies(train_df['Embarked'], prefix="Embarked")],
                     axis=1)
test_df = pd.concat([test_df, pd.get_dummies(test_df['Pclass'], 
                                             prefix="PClass"),
                      pd.get_dummies(test_df['Sex'], prefix="Sex"),
                      pd.get_dummies(test_df['SibSp'], prefix="SibSp"),
                      pd.get_dummies(test_df['Parch'], prefix="Parch"),
                    pd.get_dummies(test_df['Embarked'], prefix="Embarked")],
                     axis=1)

In [38]:
train_df.drop(['Survived', 'Pclass', 'Name', 'Sex', 'SibSp', 'Parch', 'Ticket', 'Cabin', 'Embarked', 'PassengerId'], 
              axis=1, inplace=True)
test_df.drop(['Pclass', 'Name', 'Sex', 'SibSp', 'Parch', 'Ticket', 'Cabin', 'Embarked', 'PassengerId'], 
             axis=1, inplace=True)

In [39]:
train_df.shape, test_df.shape

((891, 24), (418, 25))

In [40]:
set(test_df.columns) - set(train_df.columns)

{'Parch_9'}

In [41]:
test_df.drop(['Parch_9'], axis=1, inplace=True)

In [42]:
train_df.head()

Age     Fare  PClass_1  PClass_2  PClass_3  Sex_female  Sex_male  SibSp_0  \
0  22.0   7.2500         0         0         1           0         1        0   
1  38.0  71.2833         1         0         0           1         0        0   
2  26.0   7.9250         0         0         1           1         0        1   
3  35.0  53.1000         1         0         0           1         0        0   
4  35.0   8.0500         0         0         1           0         1        1   

   SibSp_1  SibSp_2  ...  Parch_0  Parch_1  Parch_2  Parch_3  Parch_4  \
0        1        0  ...        1        0        0        0        0   
1        1        0  ...        1        0        0        0        0   
2        0        0  ...        1        0        0        0        0   
3        1        0  ...        1        0        0        0        0   
4        0        0  ...        1        0        0        0        0   

   Parch_5  Parch_6  Embarked_C  Embarked_Q  Embarked_S  
0        0        0           0           0           1  
1        0        0           1           0           0  
2        0        0           0           0           1  
3        0        0           0           0           1  
4        0        0           0           0           1  

[5 rows x 24 columns]

In [43]:
test_df.head()

Age     Fare  PClass_1  PClass_2  PClass_3  Sex_female  Sex_male  SibSp_0  \
0  34.5   7.8292         0         0         1           0         1        1   
1  47.0   7.0000         0         0         1           1         0        0   
2  62.0   9.6875         0         1         0           0         1        1   
3  27.0   8.6625         0         0         1           0         1        1   
4  22.0  12.2875         0         0         1           1         0        0   

   SibSp_1  SibSp_2  ...  Parch_0  Parch_1  Parch_2  Parch_3  Parch_4  \
0        0        0  ...        1        0        0        0        0   
1        1        0  ...        1        0        0        0        0   
2        0        0  ...        1        0        0        0        0   
3        0        0  ...        1        0        0        0        0   
4        1        0  ...        0        1        0        0        0   

   Parch_5  Parch_6  Embarked_C  Embarked_Q  Embarked_S  
0        0        0           0           1           0  
1        0        0           0           0           1  
2        0        0           0           1           0  
3        0        0           0           0           1  
4        0        0           0           0           1  

[5 rows x 24 columns]

In [44]:
from sklearn.model_selection import train_test_split, cross_val_score
import numpy as np

In [45]:
y.shape

(891,)

In [46]:
x = train_df

In [47]:
x.shape

(891, 24)

In [48]:
first_tree = DecisionTreeClassifier(max_depth=2,random_state=17)

In [49]:
first_tree.fit(x, y)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=2, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=17, splitter='best')

In [50]:
import os

os.environ['PATH'] = os.environ['PATH']+';'+os.environ['CONDA_PREFIX']+r"\Library\bin\graphviz"

In [51]:
from ipywidgets import Image
from io import StringIO
import pydotplus
from sklearn.tree import export_graphviz

dot_data = StringIO()
export_graphviz(first_tree, feature_names=x.columns, 
                out_file=dot_data, filled=True)
graph = pydotplus.graph_from_dot_data(dot_data.getvalue())
Image(value=graph.create_png())

Image(value=b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x02\x89\x00\x00\x01g\x08\x06\x00\x00\x00\xcaq\xd3N\x…

## Сколько признаков задействуются при прогнозе деревом решений глубины 2?
* 3

In [54]:
predicted_df = first_tree.predict(test_df)

In [55]:
def write_to_submission_file(predicted_labels, out_file, train_num=891,
                    target='Survived', index_label="PassengerId"):
    # turn predictions into data frame and save as csv file
    predicted_df = pd.DataFrame(predicted_labels,
                                index = np.arange(train_num + 1,
                                                  train_num + 1 +
                                                  predicted_labels.shape[0]),
                                columns=[target])
    predicted_df.to_csv(out_file, index_label=index_label)

In [56]:
write_to_submission_file(predicted_df, 'predicted_df4.csv')

## Сформируйте файл посылки и отправьте на Kaggle
* 0.746

In [57]:
from sklearn.model_selection import GridSearchCV

In [58]:
tree_params = {'max_depth': list(range(1, 5)), 
               'min_samples_leaf': list(range(1, 5))}

In [59]:
tree_grid = GridSearchCV(first_tree, tree_params, cv=5, n_jobs=-1)

In [61]:
%%time
tree_grid.fit(x,y)

Wall time: 1.83 s


GridSearchCV(cv=5, error_score=nan,
             estimator=DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None,
                                              criterion='gini', max_depth=2,
                                              max_features=None,
                                              max_leaf_nodes=None,
                                              min_impurity_decrease=0.0,
                                              min_impurity_split=None,
                                              min_samples_leaf=1,
                                              min_samples_split=2,
                                              min_weight_fraction_leaf=0.0,
                                              presort='deprecated',
                                              random_state=17,
                                              splitter='best'),
             iid='deprecated', n_jobs=-1,
             param_grid={'max_depth': [1, 2, 3, 4],
                         'min_sam

In [63]:
tree_grid.best_params_

{'max_depth': 3, 'min_samples_leaf': 3}

In [73]:
np.mean(cross_val_score(tree_grid, x, y, cv=5))

0.8091959073504489

## Каковы лучшие параметры дерева, настроенные на кросс-валидации с помощью GridSearchCV?
* max_depth=3, min_samples_leaf=3

In [64]:
predicted_df2 = tree_grid.predict(test_df)

In [65]:
def write_to_submission_file(predicted_labels, out_file, train_num=891,
                    target='Survived', index_label="PassengerId"):
    # turn predictions into data frame and save as csv file
    predicted_df2 = pd.DataFrame(predicted_labels,
                                index = np.arange(train_num + 1,
                                                  train_num + 1 +
                                                  predicted_labels.shape[0]),
                                columns=[target])
    predicted_df2.to_csv(out_file, index_label=index_label)

In [66]:
write_to_submission_file(predicted_df2, 'predicted_df5.csv')

## Каков результат второй посылки (дерево решений с настройкой гиперпараметров) в публичном рейтинге соревнования Titanic?
* 0.7799

In [74]:
second_tree = DecisionTreeClassifier(max_depth=3, min_samples_leaf=3, random_state=17)

In [75]:
first_tree.fit(x, y)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=2, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=17, splitter='best')

In [76]:
np.mean(cross_val_score(second_tree, x, y, cv=5))

0.8103132257862031

## Какой получилась средняя доля верных ответов на кросс-валидации для дерева решений с лучшим сочетанием гиперпараметров max_depth и min_samples_leaf?
* 0.81